File: src/main/java/com/jcloisterzone/feature/Tower.java
package com.jcloisterzone.feature;

import com.jcloisterzone.board.Location;
import com.jcloisterzone.board.Position;
import com.jcloisterzone.board.Rotation;
import com.jcloisterzone.board.pointer.FeaturePointer;
import com.jcloisterzone.game.capability.TowerCapability;

import io.vavr.collection.List;

public class Tower extends TileFeature implements Structure {

    private static final List<FeaturePointer> INITIAL_PLACE = List.of(new FeaturePointer(Position.ZERO, Tower.class, Location.I));

    private List<TowerCapability.TowerToken> pieces = List.empty();

    public Tower() {
        this(INITIAL_PLACE, List.empty());
    }

    public Tower(List<FeaturePointer> places, List<TowerCapability.TowerToken> pieces) {
        super(places);
        this.pieces = pieces;
    }

    @Override
    public Tower placeOnBoard(Position pos, Rotation rot) {
        return new Tower(placeOnBoardPlaces(pos, rot), pieces);
    }

    public Tower addPiece(TowerCapability.TowerToken token) {
        return new Tower(places, pieces.append(token));
    }

    public List<TowerCapability.TowerToken> getPieces() {
        return pieces;
    }

    public boolean matchLastPiece(TowerCapability.TowerToken token) {
    	if (getPieces().size()==0) {
    		return false;
    	}
    	return getPieces().lastOption().get().equals(token);
    }

    public static String name() {
        return "Tower";
    }
}

File: src/main/ts/com/jcloisterzone/feature/Tower.ts
import { List } from "../../../io/vavr/SeqTypes.js";
import { Location } from "../board/Location.js";
import { Position } from "../board/Position.js";
import type { Rotation } from "../board/Rotation.js";
import { FeaturePointer } from "../board/pointer/FeaturePointer.js";
import type { TowerCapability } from "../game/capability/TowerCapability.js";
import type { Feature } from "./Feature.js";
import type { Structure } from "./Structure.js";
import { TileFeature } from "./TileFeature.js";

type TowerToken = TowerCapability.TowerToken;

/** A tower (The Tower expansion). Holds stacked tower pieces. */
export class Tower extends TileFeature implements Structure {
  static readonly simpleName = "Tower";

  private static readonly INITIAL_PLACE = List.of(
    new FeaturePointer(Position.ZERO, Tower, Location.I),
  );

  private readonly pieces: List<TowerToken>;

  constructor(places: List<FeaturePointer> = Tower.INITIAL_PLACE, pieces: List<TowerToken> = List.empty<TowerToken>()) {
    super(places);
    this.pieces = pieces;
  }

  placeOnBoard(pos: Position, rot: Rotation): Feature {
    return new Tower(this.placeOnBoardPlaces(pos, rot), this.pieces);
  }

  addPiece(token: TowerToken): Tower {
    return new Tower(this.places, this.pieces.append(token) as List<TowerToken>);
  }

  getPieces(): List<TowerToken> {
    return this.pieces;
  }

  matchLastPiece(token: TowerToken): boolean {
    if (this.pieces.size() === 0) {
      return false;
    }
    return this.pieces.lastOption().get() === token;
  }
}
